syz-verifier/main.go

```diff
@@ -0,0 +1,310 @@
+	}
+
+	s, err := rpctype.NewRPCServer(vrf.addr, "Verifier", srv)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Printf("serving rpc on tcp://%v", s.Addr())
+	srv.port = s.Addr().(*net.TCPAddr).Port
+
+	go s.Serve()
+	return srv, nil
+}
+
+// Connect notifies the RPCServer that a new Runner was started.
+func (srv *RPCServer) Connect(a *rpctype.RunnerConnectArgs, r *int) error {
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+	pool, vm := a.Pool, a.VM
+	srv.pools[pool].vmRunners[vm] = []int{}
+	return nil
+}
+
+// NewExchange is called when a Runner requests a new program to execute and, potentially, wants to send a new Result to the RPCServer.
+func (srv *RPCServer) NewExchange(a *rpctype.NewExchangeArgs, r *rpctype.NewExchangeRes) error {
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+
+	if a.Info.Calls != nil {
+		res := &verf.Result{
+			Pool:    a.Pool,
+			ProgIdx: a.ProgIdx,
+			Hanged:  a.Hanged,
+			Info:    a.Info,
+		}
+		if resReady := srv.newResult(res); resReady {
+			srv.sendForVerification(a.ProgIdx)
+		}
+	}
+
+	prog, pi := srv.newProgram(a.Pool, a.VM)
+	r.RPCProg = rpctype.RPCProg{Prog: prog, ProgIdx: pi}
+	return nil
+}
+
+func (srv *RPCServer) newProgram(pool int, vm int) ([]byte, int) {
+	if len(srv.pools[pool].progs) == 0 {
+		// generate new program
+		srv.progIdx++
+		vrf := srv.vrf
+		prog := vrf.target.Generate(vrf.rnd, prog.RecommendedCalls, vrf.choiceTable)
+		pi := &programInfo{
+			prog:       prog,
+			idx:        srv.progIdx,
+			serialized: prog.Serialize(),
+			res:        make([]*verf.Result, 0),
+			left:       make(map[int]bool),
+		}
+		for idx := range srv.pools {
+			srv.pools[idx].progs = append(srv.pools[idx].progs, pi.idx)
+			pi.left[idx] = true
+		}
+		srv.progs[srv.progIdx] = pi
+	}
+
+	idx := srv.pools[pool].progs[0]
+	p := srv.progs[idx]
+	srv.pools[pool].vmRunners[vm] = append(srv.pools[pool].vmRunners[vm], idx)
+	srv.pools[pool].progs = srv.pools[pool].progs[1:]
+	return p.serialized, p.idx
+}
+
+// newResult is called when a Runner sends a new Result. It returns true if all Results from the corresponding programs have been received and they can be sent for verification. Otherwise, it returns false.
+func (srv *RPCServer) newResult(res *verf.Result) bool {
+	idx := res.ProgIdx
+	srv.progs[idx].res = append(srv.progs[idx].res, res)
+	delete(srv.progs[idx].left, res.Pool)
+	return len(srv.progs[idx].left) == 0
+}
+func (srv *RPCServer) sendForVerification(progIdx int) {
+	info := srv.progs[progIdx]
+	go verf.Verify(info.res, info.prog)
+	delete(srv.progs, progIdx)
+}
+
+// cleanup is called when a vm.Instance crashes.
+func (srv *RPCServer) cleanup(poolIdx int, vmIdx int) {
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+	progs := srv.pools[poolIdx].vmRunners[vmIdx]
+	delete(srv.pools[poolIdx].vmRunners, vmIdx)
+	for _, idx := range progs {
+		if p := srv.progs[idx]; len(p.left) == 1 && p.left[poolIdx] {
+			srv.sendForVerification(idx)
+			continue
+		}
+		delete(srv.progs[idx].left, poolIdx)
```

Comment:
I think there is an issue in this logic.

Reply:
Consider if there are more than 2 pools and some subset of them crashes.

Reply:
I can't see a problem with that case, even if they will call `cleanup` concurrently, the mutex will make the function execute sequentially.

Reply:
I think I mis-read the code before (where we delete from what map). I don't see any issue with the logic now.
I think it would be easier to read as:
```
		delete(prog.left, poolIdx)
		if len(prog.left) == 0 {
			go verf.Verify(prog.res, prog.prog)
			delete(srv.progs, idx)
		}
```

Reply:
thanks for the suggestion!